### ISSUE TITLE
Panel does not correctly separate the required frontend assets of distinct served apps

### ISSUE THREAD

**Issue by apirogov** [holoviz/panel#5598]
#### ALL software version info
Panel 1.2.3

#### Description of expected behavior and the observed behavior

When I serve multiple apps, I expect that only the JS / CSS that is needed for an app is served to the frontend when a user accesses the app. However, it seems like the list on loaded/needed assets is somehow "shared" between apps.

Now e.g. if I use the CodeEditor in app A, but do not use it in app B, still the assets for CodeEditor are loaded when accessing app B.

#### Complete, minimal, self-contained example code that reproduces the issue

Launch a panel server with these two apps:

```
import panel as pn

pn.serve({
    'editor': pn.widgets.CodeEditor(filename="test.md", value="hello test"),
    'simple': pn.widgets.StaticText(value="hello world"),
}, port=1234)
```

1. access `http://localhost:1234/simple` - it looks okay (I guess).

2. access the other app using the codeeditor at `http://localhost:1234/editor` - now the ACE editor JS is loaded (as it should)

3. reload/open new tab with `http://localhost:1234/simple`. **Now the ACE Component is loaded, even though it is clearly not needed for this app**

I think this is quite problematic for use-cases like ours, where panel is used to serve a possibly large number of completely unrelated apps that can have various different widgets. It seems like in the worst case, panel will send all the JS for all of the widgets used **anywhere** in some app - which is wasteful and inefficient.

#### Screenshots or screencasts of the bug in action

'Simple' app, **before** loading 'editor app':

![2023-10-09-141357_1802x288_scrot](https://github.com/holoviz/panel/assets/371708/e021c0c6-4d87-40be-910a-5533dbf74777)

the 'editor app' assets:
![2023-10-09-141427_1482x283_scrot](https://github.com/holoviz/panel/assets/371708/bf9ecda0-d89b-4c1e-99a2-10ca8d30056c)

'Simple' app, **after** accessing the second app at least once:
![2023-10-09-141445_1392x282_scrot](https://github.com/holoviz/panel/assets/371708/bbb7e747-c64e-42d7-a380-da921b32be66)

**Comment by philippjfr** [comment]
This is correct however there is a recommended way of dealing with this, specifically if you want to load only the required extensions in each app you should load them explicitly with `pn.extension`, i.e. in your example:

```python
import panel as pn

def editor():
    pn.extension('codeeditor')
    return pn.widgets.CodeEditor(filename="test.md", value="hello test")

def simple():
    return pn.widgets.StaticText(value="hello world")

pn.serve({
    'editor': editor,
    'simple': simple,
}, port=1234)
```

**Comment by apirogov** [comment]
I tried out your example, but it still behaves the same way - open editor page, then open 'simple' page, ace.js is still loaded.

Also I don't see how calling a global method to declare that I need a certain extension would make it "locally scoped"? Unless I don't understand some magic, this information must somehow be connected to the app directly. How would panel know which apps use editor() except by source code analysis?

**Comment by philippjfr** [comment]
My apologies, my code should have read:

```python
import panel as pn

def editor():
    pn.extension('codeeditor')
    return pn.widgets.CodeEditor(filename="test.md", value="hello test")

def simple():
    pn.extension()
    return pn.widgets.StaticText(value="hello world")

pn.serve({
    'editor': editor,
    'simple': simple,
}, port=1234)
```

i.e. each app needs to either scope the extensions locally or it will inherit all globally loaded extensions.

>Unless I don't understand some magic, this information must somehow be connected to the app directly. How would panel know which apps actually use editor() (and calling editor here is the only way to activate the pn.extension statement), except by source code analysis?

Internally `pn.extension` tracks whether it is being called in global scope or in session scope. If it's in session scope it'll register the declared extensions per session while if it is called globally it'll register the extension to be included globally.

**Comment by apirogov** [comment]
Ah okay, the fixed code works, and thanks for the explanation!

You tagged this with 'docs' - I agree, it would be nice to have this info.

All I stumbled upon in that context was the general recommendation of declaring the used extensions, but I did not stumble on any scoping info for that.

I'd expect to find this discussed somewhere in the docs about deploying / embedding panel apps.

Just one remaining question - we use panel in a very dynamic way, e.g. a "dashboard" app endpoint delivers distinct widgets (i.e. assembles totally unrelated dashboards), depending on other context/parameters. You said "session scope", by that you really mean the scope of individual app *instantiations* (i.e.,  distinct accesses to the endpoint)? If this is the case, this would be/is perfect.

**Comment by philippjfr** [comment]
>You said "session scope", by that you really mean the scope of individual app instantiations (i.e., distinct accesses to the endpoint)? If this is the case, this would be/is perfect.

That's exactly right, a "session" is a specific instantiation of a particular app for a particular user which can completely change depending on external state, the user, URL parameters or anything else.
